src/symbol/projection.js

```diff
@@ -0,0 +1,247 @@
+        mat4.scale(m, m, [1 / pixelsToTileUnits, 1 / pixelsToTileUnits, 1]);
+        if (!rotateWithMap) {
+            mat4.rotateZ(m, m, transform.angle);
+        }
+    } else {
+        mat4.scale(m, m, [transform.width / 2, -transform.height / 2, 1]);
+        mat4.translate(m, m, [1, -1, 0]);
+        mat4.multiply(m, m, posMatrix);
+    }
+    return m;
+}
+
+/*
+ * Returns a matrix for converting from the correct label coordinate space to gl coords.
+ */
+function getGlCoordMatrix(posMatrix, pitchWithMap, rotateWithMap, transform, pixelsToTileUnits) {
+    const m = mat4.identity(new Float32Array(16));
+    if (pitchWithMap) {
+        mat4.multiply(m, m, posMatrix);
+        mat4.scale(m, m, [pixelsToTileUnits, pixelsToTileUnits, 1]);
+        if (!rotateWithMap) {
+            mat4.rotateZ(m, m, -transform.angle);
+        }
+    } else {
+        mat4.scale(m, m, [1, -1, 1]);
+        mat4.translate(m, m, [-1, -1, 0]);
+        mat4.scale(m, m, [2 / transform.width, 2 / transform.height, 1]);
+    }
+    return m;
+}
+
+function project(point, matrix) {
+    const pos = [point.x, point.y, 0, 1];
+    vec4.transformMat4(pos, pos, matrix);
+    return new Point(pos[0] / pos[3], pos[1] / pos[3]);
+}
+
+function isVisible(anchorPos, symbol, clippingBuffer, painter) {
+    const x = anchorPos[0] / anchorPos[3];
+    const y = anchorPos[1] / anchorPos[3];
+    const inPaddedViewport = (
+            x >= -clippingBuffer[0] &&
+            x <= clippingBuffer[0] &&
+            y >= -clippingBuffer[1] &&
+            y <= clippingBuffer[1]);
+    return inPaddedViewport && painter.frameHistory.isVisible(symbol.placementZoom);
+}
+
+/*
+ *  Update the `dynamicLayoutVertexBuffer` for the buffer with the correct glyph positions for the current map view.
+ *  This is only run on labels that are aligned with lines. Horizontal labels are handled entirely in the shader.
+ */
+function updateLineLabels(bucket, posMatrix, painter, isText, labelPlaneMatrix, pitchWithMap, keepUpright, pixelsToTileUnits, layer) {
+
+    const sizeData = isText ? bucket.textSizeData : bucket.iconSizeData;
+    const partiallyEvaluatedSize = symbolSize.evaluateSizeForZoom(sizeData, painter.transform, layer, isText);
+
+    const clippingBuffer = [256 / painter.width * 2 + 1, 256 / painter.height * 2 + 1];
+
+    const dynamicLayoutVertexArray = isText ?
+        bucket.buffers.glyph.dynamicLayoutVertexArray :
+        bucket.buffers.icon.dynamicLayoutVertexArray;
+    dynamicLayoutVertexArray.clear();
+
+    const lineVertexArray = bucket.lineVertexArray;
+    const placedSymbols = isText ? bucket.placedGlyphArray : bucket.placedIconArray;
+
+    for (let s = 0; s < placedSymbols.length; s++) {
+        const symbol = placedSymbols.get(s);
+
+        const anchorPos = [symbol.anchorX, symbol.anchorY, 0, 1];
+        vec4.transformMat4(anchorPos, anchorPos, posMatrix);
+
+        // Don't bother calculating the correct point for invisible labels. Move them offscreen.
```

Comment:
Maybe add a note here that we have to move them offscreen instead of just dropping them because we're not updating the paired/non-dynamic buffers?